### ISSUE TITLE
Dropping support for EOL PHP versions

### ISSUE THREAD

**Issue by HoldYourWaffle** [flat3/lodata#754]
Lodata v5 still supports EOL PHP versions 7.3, 7.4 and (very soon) 8.0.
Dropping support for these outdated versions would give access to modern language features that make using, maintaining and improving Lodata a lot easier.

**Dropping PHP 7.3 (EOL 2 years ago) would give access to...**
1. [Typed properties](https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.typed-properties)
2. [Arrow functions](https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.arrow-functions)
3. [Null coalescing assignment operator](https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.null-coalescing-assignment-operator)

**Dropping PHP 7.4 (EOL 1 year ago) would give access to...**
1. [Named arguments](https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments)
2. [Attributes](https://www.php.net/manual/en/language.attributes.overview.php)
3. [Constructor property promotion](https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion)
4. [Union types](https://www.php.net/manual/en/language.types.type-system.php#language.types.type-system.composite.union)
5. [`match` expression](https://www.php.net/manual/en/control-structures.match.php)
6. [Nullsafe operator](https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.nullsafe)
7. [`mixed` type](https://www.php.net/manual/en/language.types.type-system.php#language.types.type-system.alias)

**Dropping PHP 8.0 (EOL in 3 days) would give access to...**
1. [Enumerations](https://www.php.net/manual/en/language.enumerations.basics.php)
2. [First class callables](https://www.php.net/manual/en/functions.first_class_callable_syntax.php)
3. [`readonly` properties](https://www.php.net/manual/en/language.oop5.properties.php#language.oop5.properties.readonly-properties)

**Anecdotally...**
1. Earlier PHP versions seem to be extra problematic in the context of #747.

This requires a major version bump of course, which is the perfect opportunity for batching some more rigorous refactors that might involve breaking changes.
In particular I'd like to simplify and reduce coupling in the operations system (#751), and I've seen similar opportunities for improvement in other systems.

Dropping these PHP versions would also drop Laravel 8, which has been EOL for almost a year.
I think it's worth seriously considering dropping support for Laravel 9 as well, considering it'll also be EOL in 3 months. I don't know enough about the codebase (or Laravel) to know if this would bring any immediate benefits, but given that it'll be EOL in 3 months I think it'd be silly if we have to support it for another major version cycle.

I'd love to hear your thoughts on this!
Suffice to say I would love to help with the effort, I'll be using Lodata in production for the foreseeable future, so I'd love to help maintain it :)

**Comment by 27pchrisl** [comment]
Hi @HoldYourWaffle, thanks for the comments! Yes I think at this point it would be reasonable to have another major version that was PHP 8.1+ only. There's certainly enough functionality in v5 for older versions.

**Comment by HoldYourWaffle** [comment]
Hi @27pchrisl, it's been quite a while, but I'm finally able to get back on #751.
While I could probably get that working on the 5.x branch without breaking changes, starting a 6.x branch now would likely make things a lot easier, allowing for more thorough modernizing refactors.

How do you feel about this?

**Comment by 27pchrisl** [comment]
Yes I'd be fine with a 6.x branch that dropped PHP versions earlier than 8.1.

**Comment by HoldYourWaffle** [comment]
Sweet!
As far as I know I can't create a PR for a new branch, so if you could get that started (perhaps with just a message in the README) I can take it from there :)

As for more rigorous refactors, I thoroughly enjoy cleaning up, refactoring and modernizing code, but I realize a lot of people prefer a more 'timid' approach of only changing the things that absolutely need changing. What's your stance on this?
